### PR TITLE
Fix destroy unit short circuit for k8s

### PR DESCRIPTION
**Pull request by wallyworld**
Fix 2 key issues with the k8s operator and uniter workers.
The first is the permission check used for the facade authoriser. If the unit was removed, a permission denied error was returned, causing the entire operator worker to bounce and mess up unit leadership. Instead, the tag of the requested unit is checked to see if it matches the authenticated application agent. Any subsequent sate api call will return not found if the unit is accessed. The logic to do the permission checks was duplicated in a few places, so is not extracted to a common method.
Also, the Remove() api call is now a no op if the unit has already been removed.

Secondly, when a unit goes dying and is destroyed, we check to see if the unit agent had started - if not we can short circuit the cleanup and remove directly. However, if there is contention for the machine execution lock and a lot of k8s unit churn caused by many pod-spec-set calls, the unit agent may not have updated agent status even though it had started. This would lead to juju removing the unit and leaving the unit agent running. This then leads to the issue in item one, where an operation on the unit gets a permission error and the entire operator worker restarts. This causes the leadership tracker worker to flip the unit leader to a minion and messes up leadership tracking.
The issue is fixed by adjusting the initial agent state to "installing agent". This is then updated to "initialising agent" when the unit agent starts, allowing juju to know whether to short circuit a unit destroy operation or not.

## QA steps

bootstrap k8s
deploy kubeflow

## Bug reference

https://bugs.launchpad.net/bugs/1898966

### CODE REVIEW

**Review by ycliuhw — APPROVED** [review]
lgtm, QA was fine.

File: apiserver/facades/agent/uniter/uniter.go [hunk 1]
```diff
@@ -178,7 +179,7 @@ func NewUniterAPI(context facade.Context) (*UniterAPI, error) {
 		return nil, errors.Trace(err)
 	}
 
-	accessUnit := unitAccessor(authorizer, st)
+	accessUnit := caasapi.CAASUnitAccessor(authorizer, caasapi.Backend(st))
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/agent/uniter/uniter.go:182]: yeah, good point

**Comment by wallyworld** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[1898966 Leadership being revoked during hook](https://trello.com/c/Cw5QrCCD/1219-1898966-leadership-being-revoked-during-hook)

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**